src/Validation.hs

```diff
@@ -62,86 +61,202 @@ module Validation
+>>> :{
+unsafeUserFromForm :: Form -> User
+unsafeUserFromForm Form{..} = User
+    { userName     = UserName formUserName
+    , userPassword = Password formPassword
+    }
+:}
+
+However, this conversion is unsafe (as name suggests) since @Form@ can
+contain /invalid/ data. So. before creating @User@ we want to check
+whether all @Form@ fields satisfy our preconditions. Specifically:
+
+1. User name must not be empty.
+2. Password should be at least 8 characters long.
+3. Password should contain at least 1 digit.
+
+'Validation' offers __modular__ and __composable__ way of defining and
+outputting all validation failures.
+
+It's a good practice to define all possible errors as a single sum
+type, so let's do this:
 
 >>> :{
-data Computer = Computer
-    { computerRam  :: !Int  -- ^ Ram in Gigabytes
-    , computerCpus :: !Int
-    } deriving stock (Eq, Show)
+data FormValidationError
+    = EmptyName
+    | ShortPassword
+    | NoDigitPassword
+    deriving stock (Show)
 :}
 
-You can validate that the computer has a minimum of 16GB of RAM:
+With 'Validation' we can define checks for separate fields
+independently and compose them later. First, let's start with defining
+validation for the name:
 
 >>> :{
-validateRam :: Int -> Validation [String] Int
-validateRam ram
-    | ram >= 16 = Success ram
-    | otherwise = Failure ["Not enough RAM"]
+validateName :: String -> Validation (NonEmpty FormValidationError) UserName
+validateName name
+    | null name = Failure (EmptyName :| [])
+    | otherwise = Success (UserName name)
 :}
 
-and that the processor has at least two CPUs:
+You can notice a few things about this function:
+
+1. All errors are collected in @NonEmpty@, since we want to have
+guarantees that in case of errors we have at least one failure.
+2. It wraps the result into @UserName@ to tell that validation is
+passed.
+
+Let's see how this function works:
+
+>>> validateName "John"
+Success "John"
+>>> validateName ""
+Failure (EmptyName :| [])
+
+Now we can define all checks for the password separately:
 
 >>> :{
-validateCpus :: Int -> Validation [String] Int
-validateCpus cpus
-    | cpus >= 2 = Success cpus
-    | otherwise = Failure ["Not enough CPUs"]
+validateShortPassword :: String -> Validation (NonEmpty FormValidationError) Password
+validateShortPassword password
+    | length password < 8 = Failure (ShortPassword :| [])
+    | otherwise = Success (Password password)
 :}
 
-You can use these functions with the 'Applicative' instance of the 'Validation'
-type to construct a validated @Computer@. You will get either (pun intended) a
-valid @Computer@ or the errors that prevent it from being considered valid.
+>>> :{
+validatePasswordDigit :: String -> Validation (NonEmpty FormValidationError) Password
+validatePasswordDigit password
+    | any isDigit password = Success (Password password)
+    | otherwise = Failure (NoDigitPassword :| [])
+:}
 
-Like so:
+Now we can easily compose those two checks to validate password:
```

Comment:
```suggestion
Now we can easily compose those two checks to validate a password:
```

Reply:
Maybe a few words on why it is easily and what validation instances and properties are used in the combination?